database.cc

```diff
@@ -2268,6 +2268,69 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
     });
 }
 
+future<std::unordered_map<sstring, std::unordered_map<sstring, db::snapshot_ctl::snapshot_details>>> database::get_snapshot_details() {
+    return seastar::async([this] {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        auto dirs_only_entries_ptr =
+            make_lw_shared<lister::dir_entry_types>(lister::dir_entry_types{directory_entry_type::directory});
+        std::unordered_map<sstring, std::unordered_map<sstring, snapshot_ctl::snapshot_details>> details;
+
+        for (auto& datadir : data_dirs) {
+            lister::scan_dir(datadir, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details] (fs::path parent_dir, directory_entry de) mutable {
+                // KS directory
+                sstring ks_name = de.name;
+
+                return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable {
+                    // CF directory
+                    sstring cf_name = de.name.substr(0, de.name.find("-"));
```

Comment:
Do we have to open-code this filename parsing here - don't we already have it in the sstable code and can call that?
I'm worried that the directory structure or filename structure might change between sstable versions. In fact, didn't it change since KA? Or maybe we don't care about supporting KA any more?

Reply:
I wasn't aware of any function that would already provide cf_name <-> cf_dir_name translation, but I'll look some more.

Reply:
We have a mapping from a sstable file name to the sstable attributes (version, generation, etc.) in the sstable class but I don't remember having something like that for tables.

There's `keyspace::column_family_directory` that maps `<cf_name, uuid>` into the directory entry name
but I saw no reverse mapping of that.
Maybe the mapping should be implemented, along side ``keyspace::column_family_directory` so it would be easier to change them together if and when this happens.